src/abacus/metrics.cpp

```diff
@@ -82,29 +71,70 @@ auto metrics::metric_value(std::size_t index) const -> uint64_t
     return *detail::raw_value(m_data, index);
 }
 
-auto metrics::initialize_metric(std::size_t index, const std::string& name)
-    -> metric
+auto metrics::metric_index(const std::string& name) const -> std::size_t
 {
-    assert(!is_metric_initialized(index));
-
-    // We check for less than since there also needs to be at least one
-    // zero terminating byte
     assert(name.size() < m_max_name_bytes);
 
-    char* name_data = detail::raw_name(m_data, index);
-    uint64_t* value_data = detail::raw_value(m_data, index);
+    for (std::size_t i = 0; i < m_metrics_count; ++i)
+    {
+        if (is_metric_initialized(i) && metric_name(i) == name)
+        {
+            return i;
+        }
+    }
+
+    assert(false && "Metric index was not found");
+}
+
+auto metrics::scope() const -> std::string
+{
+    return m_scope;
+}
+
+auto metrics::initialize_metric(const std::string& name) -> metric
+{
+    std::string scoped_name = m_scope + "." + name;
+
+    assert(scoped_name.size() < m_max_name_bytes);
+
+    char* name_data = detail::raw_name(m_data, m_metrics_count);
+    uint64_t* value_data = detail::raw_value(m_data, m_metrics_count);
 
     // Copy the name
-    std::memcpy(name_data, name.data(), name.size());
+    std::memcpy(name_data, scoped_name.data(), scoped_name.size());
 
     // Use memcpy here for now, since placement new causes Bus Error on
     // Raspberry pi
     uint64_t value = 0U;
     std::memcpy(value_data, &value, sizeof(uint64_t));
 
+    m_metrics_count++;
+
     return metric{value_data};
 }
 
+auto metrics::metrics_count() const -> std::size_t
+{
+    return m_metrics_count;
+}
+
+void metrics::add_scope(const std::string& text)
```

Comment:
This seems dangerous, when you chose the max size of the names you may not know the number of scopes to be added later?